src/libraries/System.Text.RegularExpressions/gen/RegexGenerator.Emitter.cs

```diff
@@ -335,6 +382,65 @@ private static void AddIsECMABoundaryHelper(Dictionary<string, string[]> require
+                {
+                    "/// <summary>A node of the Aho-Corasick algorithm's trie data structure. Each node represents",
+                    "/// a word that is formed from the path of characters that leads from the root node to this one.</summary>",
+                    "internal readonly struct TrieNode",
+                    "{",
+                    "    /// <summary>A cached empty trie children dictionary.</summary>",
+                    "    internal static readonly Dictionary<char, int> s_EmptyChildrenDictionary = new Dictionary<char, int>();",
+                    "    /// <summary>An associative collection of characters and the node index they lead to.</summary>",
+                    "    internal Dictionary<char, int> Children { get; init; }",
+                    "    /// <summary>The index of the node the algorithm will jump to if it doesn't find a match in the node it is.</summary>",
+                    "    internal int SuffixLink { get; init; }",
+                    "    /// <summary>The length of the word that matches at this node, or -1 if there isn't such word.</summary>",
+                    "    internal int MatchLength { get; init; }",
+                    "}",
+                    "",
+                    "/// <summary>Finds the index of the longest leftmost match of one of <paramref name=\"trie\"/>'s",
+                    "/// strings in <paramref name=\"text\"/>, using the Aho-Corasick algorithm.</summary>",
+                    "internal static int TrieSearch(ReadOnlySpan<TrieNode> trie, ReadOnlySpan<char> text)",
+                    "{",
+                    "    int currentState = 0;",
+                    "",
+                    "    for (int i = 0; i < text.Length; i++)",
+                    "    {",
+                    "        char c = text[i];",
+                    "        while (true)",
+                    "        {",
+                    "            if (trie[currentState].Children.TryGetValue(c, out int nextState))",
+                    "            {",
+                    "                currentState = nextState;",
+                    "                break;",
+                    "            }",
+                    "            if (currentState == 0)",
+                    "            {",
+                    "                break;",
+                    "            }",
+                    "",
+                    "            currentState = trie[currentState].SuffixLink;",
+                    "        }",
+                    "",
+                    "        int matchLength = trie[currentState].MatchLength;",
+                    "",
+                    "        if (matchLength != -1)",
+                    "        {",
+                    "            return i + 1 - matchLength;",
+                    "        }",
+                    "    }",
+                    "",
+                    "    return -1;",
+                    "}"
```

Comment:
This approach still requires the compiled code to effectively interpret the trie, reading each character, looking up the next node, etc.  An alternative is to actually compile the trie logic down into code, emitting custom code for each state, with each state having its own label that can be jumped to.  Did you consider something like that?  Imagine instead of `if (trie[currentState].Children.TryGetValue(c, out int nextState))`, you have code for the dedicated `currentState` that contains a `switch (c)`, with a case for everything in the trie, each hardcoded with a goto for the right next state, etc. It would be optimized for the exact trie in use, it would avoid the need for the boilerplate helpers, etc., and the built-in limitations on the number of nodes in the trie should limit code explosion.

Reply:
Yes, I thought of this but left it for another time. I will give it a try.

Reply:
I did it for `Compiled` regexes and will do it in the source generator soon.

Reply:
Done. I will benchmark it tomorrow.